test/Core.Test/Auth/Services/AuthRequestServiceTests.cs

```diff
@@ -0,0 +1,309 @@
+        {
+            Key = "test_key",
+            DeviceIdentifier = "test_identifier",
+            RequestApproved = false,
+            MasterPasswordHash = "my_hash",
+        };
+
+        var udpatedAuthRequest = await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel);
+
+        Assert.Equal(udpatedAuthRequest.MasterPasswordHash, authRequest.MasterPasswordHash);
+
+        await sutProvider.GetDependency<IAuthRequestRepository>()
+            .Received()
+            .ReplaceAsync(udpatedAuthRequest);
+
+        await sutProvider.GetDependency<IPushNotificationService>()
+            .DidNotReceiveWithAnyArgs()
+            .PushAuthRequestResponseAsync(udpatedAuthRequest);
+    }
+
+    [Theory, BitAutoData]
+    public async Task UpdateAuthRequestAsync_InvalidUser_ThrowsNotFound(
+        SutProvider<AuthRequestService> sutProvider,
+        AuthRequest authRequest,
+        Guid userId)
+    {
+        // Give it a recent creation date so that it is valid
+        authRequest.CreationDate = DateTime.UtcNow.AddMinutes(-10);
+        authRequest.Approved = false;
+
+        // Auth request should not be null
+        sutProvider.GetDependency<IAuthRequestRepository>()
+            .GetByIdAsync(authRequest.Id)
+            .Returns(authRequest);
+
+        var updateModel = new AuthRequestUpdateRequestModel
+        {
+            Key = "test_key",
+            DeviceIdentifier = "test_identifier",
+            RequestApproved = true,
+            MasterPasswordHash = "my_hash",
+        };
+
+        // Give it a randomly generated userId such that it won't be valid for the AuthRequest
+        await Assert.ThrowsAsync<NotFoundException>(
+            async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, userId, updateModel));
+    }
+
+    [Theory, BitAutoData]
+    public async Task UpdateAuthRequestAsync_OldAuthRequest_ThrowsNotFound(
+        SutProvider<AuthRequestService> sutProvider,
+        AuthRequest authRequest)
+    {
+        // AuthRequest's have a valid lifetime of only 15 minutes, make it older than that
+        authRequest.CreationDate = DateTime.UtcNow.AddMinutes(-16);
+        authRequest.Approved = false;
+
+        // Auth request should not be null
+        sutProvider.GetDependency<IAuthRequestRepository>()
+            .GetByIdAsync(authRequest.Id)
+            .Returns(authRequest);
+
+        var updateModel = new AuthRequestUpdateRequestModel
+        {
+            Key = "test_key",
+            DeviceIdentifier = "test_identifier",
+            RequestApproved = true,
+            MasterPasswordHash = "my_hash",
+        };
+
+        // Give it a randomly generated userId such that it won't be valid for the AuthRequest
+        await Assert.ThrowsAsync<NotFoundException>(
+            async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel));
+    }
```

Comment:
Here's two additional test cases for the `UpdateAuthRequestAsync` method (untested):

(1) Test case for when _deviceRepository.GetByIdentifierAsync() returns null:
```
[Theory, BitAutoData]
public async Task UpdateAuthRequestAsync_InvalidDeviceIdentifier_ThrowsBadRequest(
    SutProvider<AuthRequestService> sutProvider,
    AuthRequest authRequest)
{
    authRequest.CreationDate = DateTime.UtcNow.AddMinutes(-10);
    authRequest.Approved = null;

    sutProvider.GetDependency<IAuthRequestRepository>()
        .GetByIdAsync(authRequest.Id)
        .Returns(authRequest);

    sutProvider.GetDependency<IDeviceRepository>()
        .GetByIdentifierAsync(Arg.Any<string>(), authRequest.UserId)
        .Returns((Device?)null);

    var updateModel = new AuthRequestUpdateRequestModel
    {
        Key = "test_key",
        DeviceIdentifier = "invalid_identifier",
        RequestApproved = true,
        MasterPasswordHash = "my_hash",
    };

    await Assert.ThrowsAsync<BadRequestException>(
        async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel));
}

```

(2) Test case for when the `AuthRequest` has already been approved or rejected:
```
[Theory, BitAutoData]
public async Task UpdateAuthRequestAsync_AlreadyApprovedOrRejected_ThrowsDuplicateAuthRequestException(
    SutProvider<AuthRequestService> sutProvider,
    AuthRequest authRequest)
{
    // Set CreationDate to a valid recent value and Approved to a non-null value
    authRequest.CreationDate = DateTime.UtcNow.AddMinutes(-10);
    authRequest.Approved = true;

    sutProvider.GetDependency<IAuthRequestRepository>()
        .GetByIdAsync(authRequest.Id)
        .Returns(authRequest);

    var device = new Device
    {
        Id = Guid.NewGuid(),
        Identifier = "test_identifier",
    };

    sutProvider.GetDependency<IDeviceRepository>()
        .GetByIdentifierAsync(device.Identifier, authRequest.UserId)
        .Returns(device);

    var updateModel = new AuthRequestUpdateRequestModel
    {
        Key = "test_key",
        DeviceIdentifier = "test_identifier",
        RequestApproved = true,
        MasterPasswordHash = "my_hash",
    };

    await Assert.ThrowsAsync<DuplicateAuthRequestException>(
        async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel));
}
```

Reply:
All 3 of these tests were great, I literally only had to bring in a namespace for `DuplicateAuthRequestException` and they all worked perfectly.

Reply:
Awesome!